CMakeLists.txt

```diff
@@ -88,14 +88,14 @@ if(CMAKE_C_COMPILER_ID MATCHES "GNU|AppleClang|Clang")
   # These two flags generate too many errors currently, but we
   # probably want these optimizations enabled.
 
-  # TODO: -fdelete-null-pointer-checks -Wnull-dereference
-
   # This flag avoid some subtle bugs related to standard conversions,
   # but currently does not compile because we are using too many
   # implicit conversions that potentially lose precision.
 
   # TODO: -Wconversions
-  add_compile_options(-Wempty-body -Wvla)
+  add_compile_options(
+    -Wempty-body -Wvla -Wnonnull
+    -fdelete-null-pointer-checks -Wnull-dereference)
```

Comment:
I am not sure I understand why we are explicitly enabling optimizations like `-fdelete-null-pointer-checks` in debug builds? First, why do we care about the speed of executions? Second, won't this make it harder to run the code in a debugger?

Reply:
Performance is not the reason for adding that flag. The main reason the flag is added is because `-Wnull-dereference` is only active if `-fdelete-null-pointer-checks` is enabled.

In addition, there is an obvious problem with the code if we have to check for null of a pointer after we have de-referenced the code and it's better if we catch this early. As a side node, deleting null pointer checks have the potential of removing quite large swaths of code so I see no reason to not have it enabled if we can.